### PR TITLE
feat(sampling): Implement sampling for NEC vector accelerators

### LINKED ISSUE #253: NEC Support Part 2: Sampling
As far as my research goes, NEC does not have dedicated sampling infrastructure in the same vein as contemporary general purpose processors.

But as #237  mentions it, you should be able to read all registers of the accelerator side from the host side. This includes the instruction counter (IC), which is only a little bit of ELF magic away from giving us the name of the function.

If I'm interpreting the `IC`[ in this file](https://github.com/veos-sxarr-NEC/ve_drv-kmod/blob/57db60178644c12a9905db530b6d96a20f9fcf2f/tools/monc.sh) correctly, then you should be able to read the instruction counter using memory-mapped I/O via PCI. However, the rest is pretty cryptic.

### PR DESCRIPTION
**Pull request by cvonelm**
This PR implements sampling for NEC Vector accelerators, such as the NEC SX-Aurora Tsubasa cards.

The "--nec" enables NEC sampling. This records the activity of every thread on every card, regardles of the overall sampling mode.

"--nec-readout-interval" controls the interval of the readouts

This fixes #253

### CODE REVIEW

File: src/monitor/nec_monitor_main.cpp [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+        std::vector<Thread> threads;
+
+        while (task_stream.good())
+        {
+            pid_t pid;
+            task_stream >> pid;
+            threads.emplace_back(Thread(pid));
+        }
+
+        threads.pop_back();
```
Comment (bmario) [inline src/monitor/nec_monitor_main.cpp:51]: Add comment, why it is necessary to drop the last entry

**Review by bmario — CHANGES_REQUESTED** [review]

File: src/monitor/main_monitor.cpp [hunk 2]
```diff
@@ -116,6 +116,25 @@ MainMonitor::MainMonitor() : trace_(), metrics_(trace_)
...
+    for (auto& dir_entry : std::filesystem::directory_iterator("/sys/class/ve"))
+    {
+        std::smatch nec_match;
+
+        auto path = dir_entry.path().string();
+        if (std::regex_match(path, nec_match, nec_regex))
+        {
+            auto monitor = nec_monitors_.emplace_back(
+                std::make_unique<NecMonitorMain>(trace_, std::stoi(nec_match[1])));
+            std::asssert(monitor.first);
```
Comment (bmario) [inline src/monitor/main_monitor.cpp:132]: `std::asSsert`? This cannot have been compiled successfully, ever.

File: src/monitor/main_monitor.cpp [hunk 3]
```diff
@@ -116,6 +116,25 @@ MainMonitor::MainMonitor() : trace_(), metrics_(trace_)
...
+
+    for (auto& dir_entry : std::filesystem::directory_iterator("/sys/class/ve"))
+    {
+        std::smatch nec_match;
+
+        auto path = dir_entry.path().string();
+        if (std::regex_match(path, nec_match, nec_regex))
+        {
+            auto monitor = nec_monitors_.emplace_back(
+                std::make_unique<NecMonitorMain>(trace_, std::stoi(nec_match[1])));
```
Comment (bmario) [inline src/monitor/main_monitor.cpp:131]: This code does not belong here. This should be part of the `Topology`

File: src/monitor/main_monitor.cpp [hunk 4]
```diff
@@ -167,6 +186,16 @@ MainMonitor::~MainMonitor()
         }
     }
 
+#ifdef HAVE_VEOSINFO
+    if (!nec_monitors_.empty())
```
Comment (bmario) [inline src/monitor/main_monitor.cpp:190]: This check isn't necessary

File: src/monitor/nec_thread_monitor.cpp [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+void NecThreadMonitor::monitor()
+{
+    static int reg[] = { IC };
+    uint64_t val;
+
+    ve_get_regvals(device_, nec_thread_.as_pid_t(), 1, reg, &val);
+    otf2::chrono::time_point tp = lo2s::time::now();
+    otf2_writer_.write_calling_context_sample(tp, cctx_manager_.sample_ref(val), 2,
+                                              trace_.interrupt_generator().ref());
+    std::this_thread::sleep_for(nec_readout_interval_);
```
Comment (bmario) [inline src/monitor/nec_thread_monitor.cpp:60]: Why is this here? `monitor` was always a read-out method.

File: src/monitor/nec_thread_monitor.cpp [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+
+void NecThreadMonitor::monitor()
+{
+    static int reg[] = { IC };
+    uint64_t val;
+
+    ve_get_regvals(device_, nec_thread_.as_pid_t(), 1, reg, &val);
+    otf2::chrono::time_point tp = lo2s::time::now();
+    otf2_writer_.write_calling_context_sample(tp, cctx_manager_.sample_ref(val), 2,
+                                              trace_.interrupt_generator().ref());
```
Comment (bmario) [inline src/monitor/nec_thread_monitor.cpp:59]: NEC card samples should use its own interrupt generator

File: src/util.cpp [hunk 7]
```diff
@@ -327,4 +327,42 @@ std::set<std::uint32_t> parse_list_from_file(std::filesystem::path file)
...
+        args.emplace_back(std::string(cmdline_c_str));
+        cmdline_c_str += args.back().length() + 1;
+    }
+    return args;
+}
+
+std::string get_nec_thread_comm(Thread thread)
+{
+    // For normal processes, /proc/{PID}/comm contains the name of the program running in that
+    // process For NEC processes, it instead contains the name of the program loader, ve_exec So
```
Comment (bmario) [inline src/util.cpp:350]:
```suggestion
    // process. For NEC processes, it instead contains the name of the program loader, ve_exec. So
```

File: src/util.cpp [hunk 8]
```diff
@@ -327,4 +327,42 @@ std::set<std::uint32_t> parse_list_from_file(std::filesystem::path file)
...
+    // instead, we have to extract the name of the actual NEC program we are executing from the
+    // commandline arguments to ve_exec
+    std::string thread_name = fmt::format("{}", thread);
+
+    auto args = get_thread_cmdline(thread);
+    for (std::size_t i = 0; i < args.size(); i++)
+    {
+        if (args[i] == "--")
```
Comment (bmario) [inline src/util.cpp:358]: Scanning the command-line seems pretty fragile to me. Isn't there a better way? Maybe an env variable or something?

File: CMakeLists.txt [hunk 9]
```diff
@@ -122,7 +123,8 @@ CMAKE_DEPENDENT_OPTION(USE_SENSORS "Use the libsensors to read system metrics."
 add_feature_info("USE_SENSORS" USE_SENSORS "Use the libsensors to read system metrics.")
 CMAKE_DEPENDENT_OPTION(USE_LIBAUDIT "Use libaudit for syscall name resolution" ON Audit_FOUND OFF)
 add_feature_info("USE_LIBAUDIT" USE_LIBAUDIT "Use libaudit for syscall name resolution.")
-
+CMAKE_DEPENDENT_OPTION(USE_VEOSINFO "Use veosinfo to sample NEC SX-Aurora Tsubasa cards." ON "Veosinfo_FOUND" OFF)
+add_feature_info("USE_VEOSINFO" USE_VEOSINFO "Use veosinfo to sample NEC SX-Aurora  Tsubasa cards.")
```
Comment (bmario) [inline CMakeLists.txt:127]:
```suggestion
CMAKE_DEPENDENT_OPTION(USE_VEOSINFO "Use libveosinfo to sample NEC SX-Aurora Tsubasa cards." ON "Veosinfo_FOUND" OFF)
add_feature_info("USE_VEOSINFO" USE_VEOSINFO "Use libveosinfo to sample NEC SX-Aurora Tsubasa cards.")
```

File: include/lo2s/monitor/main_monitor.hpp [hunk 10]
```diff
@@ -82,6 +84,9 @@ class MainMonitor
 #ifdef HAVE_SENSORS
     std::unique_ptr<metric::sensors::Recorder> sensors_recorder_;
 #endif
+#ifdef HAVE_VEOSINFO
+    std::vector<std::unique_ptr<NecMonitorMain>> nec_monitors_;
```
Comment (bmario) [inline include/lo2s/monitor/main_monitor.hpp:88]: Why unique_ptr in vector?

File: include/lo2s/monitor/nec_thread_monitor.hpp [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+
+protected:
+    std::string group() const override
+    {
+        return "nec::ThreadMonitor";
+    }
+
+    void finalize_thread() override;
+
+    void monitor([[maybe_unused]] int fd) override;
```
Comment (bmario) [inline include/lo2s/monitor/nec_thread_monitor.hpp:44]: I'd move `[[maybe_unused]]` to the definition of the function.

File: include/lo2s/monitor/nec_monitor_main.hpp [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+    void run() override;
+    void finalize_thread() override;
+
+private:
+    int get_nec_device_id_for_thread(Thread thread);
+
+    std::map<Thread, NecThreadMonitor> monitors_;
+    trace::Trace& trace_;
+    int device_;
+    bool stopped_;
```
Comment (bmario) [inline include/lo2s/monitor/nec_monitor_main.hpp:60]: I'm pretty sure, `stopped_` should be a `std::atomic<bool>`.

File: include/lo2s/config.hpp [hunk 13]
```diff
@@ -92,6 +92,9 @@ struct Config
     // syscalls
     bool use_syscalls = false;
     std::vector<int64_t> syscall_filter;
+    // NEC SX-Aurora Tsubasa
+    bool use_nec;
+    std::chrono::nanoseconds nec_readout_interval;
```
Comment (bmario) [inline include/lo2s/config.hpp:97]: is there a reason to call it readOUT_interval instead of read_interval like the others?

File: src/config.cpp [hunk 14]
```diff
@@ -333,6 +334,8 @@ void parse_program_options(int argc, const char** argv)
     io_options.toggle("block-io",
                       "Enable recording of block I/O events (requires access to debugfs)");
 
+    nec_options.toggle("nec", "Enable NEC SX-Aurora Tsubasa sampling");
+    nec_options.option("nec-readout-interval").optional().metavar("MSEC").default_value("1");
```
Comment (bmario) [inline src/config.cpp:338]: I don't think it's a good idea to have 1 millisecond as the default and minimal value. Why not read the interval as microseconds?

File: include/lo2s/monitor/nec_monitor_main.hpp [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+#include <lo2s/monitor/nec_thread_monitor.hpp>
+#include <lo2s/monitor/threaded_monitor.hpp>
+#include <lo2s/trace/trace.hpp>
+#include <lo2s/types.hpp>
+
+#include <veosinfo/veosinfo.h>
+
+#include <libved.h>
+
+#include <filesystem>
```
Comment (bmario) [inline include/lo2s/monitor/nec_monitor_main.hpp:33]:
```suggestion
#include <filesystem>

```

File: include/lo2s/monitor/nec_monitor_main.hpp [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+namespace lo2s
+{
+namespace monitor
+{
+class NecMonitorMain : public ThreadedMonitor
```
Comment (bmario) [inline include/lo2s/monitor/nec_monitor_main.hpp:38]: This should live in a `nec` namespace similar to `perf`.

File: include/lo2s/monitor/nec_thread_monitor.hpp [hunk 17]
```diff
@@ -0,0 +1,54 @@
...
+#pragma once
+
+#include <lo2s/monitor/poll_monitor.hpp>
+#include <lo2s/perf/calling_context_manager.hpp>
+#include <lo2s/trace/trace.hpp>
+namespace lo2s
+{
+namespace monitor
+{
+class NecThreadMonitor : public PollMonitor
```
Comment (bmario) [inline include/lo2s/monitor/nec_thread_monitor.hpp:31]: This should live in a `nec` namespace similar to `perf`.

File: include/lo2s/monitor/nec_thread_monitor.hpp [hunk 18]
```diff
@@ -0,0 +1,54 @@
...
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with lo2s.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include <lo2s/monitor/poll_monitor.hpp>
+#include <lo2s/perf/calling_context_manager.hpp>
+#include <lo2s/trace/trace.hpp>
```
Comment (bmario) [inline include/lo2s/monitor/nec_thread_monitor.hpp:26]:
```suggestion
#include <lo2s/trace/trace.hpp>

```

File: src/monitor/nec_monitor_main.cpp [hunk 19]
```diff
@@ -0,0 +1,125 @@
...
+ */
+
+#include <lo2s/monitor/nec_monitor_main.hpp>
+
+namespace lo2s
+{
+namespace monitor
+{
+
+int NecMonitorMain::get_nec_device_id_for_thread(Thread thread)
```
Comment (bmario) [inline src/monitor/nec_monitor_main.cpp:29]:
```suggestion
int NecMonitorMain::get_device_of(Thread thread)
```
We no smurfs.

File: src/monitor/nec_monitor_main.cpp [hunk 20]
```diff
@@ -0,0 +1,125 @@
...
+            break;
+        }
+    }
+    return real_device_id;
+}
+NecMonitorMain::NecMonitorMain(trace::Trace& trace, int device)
+: ThreadedMonitor(trace, fmt::format("VE {}", device)), trace_(trace), device_(device),
+  stopped_(false)
+{
+    ve_node_info(&nodeinfo_);
```
Comment (bmario) [inline src/monitor/nec_monitor_main.cpp:49]: check return for errors. It also sets errno: https://github.com/veos-sxarr-NEC/veosinfo_source/blob/master/veosinfo.c

File: src/monitor/nec_monitor_main.cpp [hunk 21]
```diff
@@ -0,0 +1,125 @@
...
+        while (task_stream.good())
+        {
+            pid_t pid;
+            task_stream >> pid;
+            threads.emplace_back(Thread(pid));
+        }
+
+        // For some god-forsaken reason, the last read entry from task_id_all will always be invalid
+
+        threads.pop_back();
```
Comment (bmario) [inline src/monitor/nec_monitor_main.cpp:69]: This should be it's own method.

File: src/monitor/nec_monitor_main.cpp [hunk 22]
```diff
@@ -0,0 +1,125 @@
...
+            }
+
+            auto ret = monitors_.emplace(std::piecewise_construct, std::forward_as_tuple(thread),
+                                         std::forward_as_tuple(thread, trace_, real_device_id));
+            if (ret.second)
+            {
+                ret.first->second.start();
+            }
+        }
+        std::this_thread::sleep_for(std::chrono::milliseconds(100));
```
Comment (bmario) [inline src/monitor/nec_monitor_main.cpp:107]: hardcoded value.

File: include/lo2s/monitor/nec_monitor_main.hpp [hunk 23]
```diff
@@ -0,0 +1,64 @@
...
+    void finalize_thread() override;
+
+private:
+    int get_nec_device_id_for_thread(Thread thread);
+
+    std::map<Thread, NecThreadMonitor> monitors_;
+    trace::Trace& trace_;
+    int device_;
+    bool stopped_;
+    struct ve_nodeinfo nodeinfo_;
```
Comment (bmario) [inline include/lo2s/monitor/nec_monitor_main.hpp:61]:
```suggestion
    ve_nodeinfo nodeinfo_;
```

**Review by bmario — CHANGES_REQUESTED** [review]

File: include/lo2s/monitor/main_monitor.hpp (hunk 10 above)
Comment (cvonelm) [reply include/lo2s/monitor/main_monitor.hpp:88]: For the same reason we have tracepoint_monitors_ in std::vector<std::unique_ptr>. NecMonitorMain's have no default constructor

File: src/monitor/nec_monitor_main.cpp [hunk 24]
```diff
@@ -0,0 +1,125 @@
...
+    for (int i = 0; i < nodeinfo_.total_node_count; i++)
+    {
+        if (!ve_check_pid(nodeinfo_.nodeid[i], thread.as_pid_t()))
+        {
+            real_device_id = nodeinfo_.nodeid[i];
+            break;
+        }
+    }
+    return real_device_id;
+}
```
Comment (bmario) [inline src/monitor/nec_monitor_main.cpp:44]: \n

File: src/monitor/nec_monitor_main.cpp [hunk 25]
```diff
@@ -0,0 +1,125 @@
...
+        std::vector<Thread> threads;
+
+        while (task_stream.good())
+        {
+            pid_t pid;
+            task_stream >> pid;
+            threads.emplace_back(Thread(pid));
+        }
+
+        //For some god-forsaken reason, the last read entry from task_id_all will always be invalid
```
[the remark this answers is not in the document]
Comment (cvonelm) [reply src/monitor/nec_monitor_main.cpp]: That is a good suggestion, but it was just a simple off-by-one error. By the time we check the stream to see if reading the pid succeeded, we have already emplaced the pid.
